test/gtest/ucs/test_conn_match.cc

```diff
@@ -0,0 +1,254 @@
+        EXPECT_EQ(conn_sn, conn_elem->conn_sn);
+        EXPECT_TRUE(!memcmp(dest_address, conn_elem->dest_address,
+                            m_conn_match_ctx.address_length));
+        return conn_elem;
+    }
+
+    void remove_conn(conn_elem_t &elem) {
+        ucs_conn_match_remove_elem(&m_conn_match_ctx, &elem.conn_match,
+                                   elem.queue_type);
+    }
+
+    ucs_conn_sn_t get_next_sn(const void *dest_address) {
+        return ucs_conn_match_get_next_sn(&m_conn_match_ctx, dest_address);
+    }
+
+private:
+    static ucs_conn_match_ctx_t m_conn_match_ctx;
+};
+
+
+ucs_conn_match_ctx_t test_conn_match::m_conn_match_ctx = {};
+
+
+class address_generator {
+public:
+    address_generator() {
+        m_seed = static_cast<uint64_t>(time(NULL));
+    }
+
+    virtual ~address_generator() {
+    }
+
+    virtual const void*  generate() = 0;
+    virtual       size_t length()   = 0;
+
+protected:
+    uint64_t m_seed;
+};
+
+
+class uuid_generator : public address_generator {
+public:
+    uuid_generator() {
+        m_current_uuid = 0;
+    }
+
+    const void *generate() {
+        m_current_uuid = ucs_generate_uuid(m_seed);
+        return &m_current_uuid;
+    }
+
+    size_t length() {
+        return sizeof(m_current_uuid);
+    }
+
+private:
+    uint64_t m_current_uuid;
+};
+
+
+class string_generator : public address_generator {
+public:
+    string_generator() {
+        m_current_str = "";
+    }
+
+    const void *generate() {
+        std::string uuid_str = ucs::to_string(ucs_generate_uuid(m_seed));
+
+        m_current_str.clear();
+        ucs::generate_random_str(m_current_str, m_length - uuid_str.size());
+        /* to make sure that the string is unique */
+        m_current_str += uuid_str;
+
+        return m_current_str.c_str();
+    }
+
+    size_t length() {
+        return m_length;
+    }
+
+private:
+    static const size_t m_length;
+    std::string  m_current_str;
+};
+
+
+const size_t string_generator::m_length = 2048 / ucs::test_time_multiplier();
+
+
+UCS_TEST_F(test_conn_match, random_insert_retrieve) {
+    const size_t        max_addresses = 128;
+    const ucs_conn_sn_t max_conns     = 128;
+    const ucs_conn_sn_t min_conns     = 1;  
+
+    std::vector<address_generator*> address_generators;
+
+    address_generators.push_back(new uuid_generator);
+    address_generators.push_back(new string_generator);
+
+    for (std::vector<address_generator*>::const_iterator it =
+             address_generators.begin(); it != address_generators.end(); it++) {
+        address_generator *generator = *it;
+        std::vector<std::vector<conn_elem_t> > elems(max_addresses);
+
+        conn_match_init(generator->length());
+
+        for (size_t i = 0; i < max_addresses; i++) {
+            const void *dest_address = generator->generate();
+
+            ucs_conn_sn_t num_conns =
+                ucs::rand() % (max_conns + 1 - min_conns) + min_conns;
```

Comment:
calculation looks weird, given that `min_conns` is always 1

Reply:
> calculation looks weird, given that `min_conns` is always 1

why weird? what if `min_conns` will be changed to `10`. then this calculation makes sense

Reply:
but min-conns is always 1?

Reply:
> but min-conns is always 1?

yes, I wanted to make it extendable in the future, if someone changed it to `10`
wdyt?

if not have common code for all values:
it will be as below to have values from `[1, max_conn]`:
```
ucs_conn_sn_t num_conns = ucs::rand() % max_conns + 1;
```

Reply:
up to you, imo better update when want to change `min_conn`

Reply:
done